### DISCUSSION TITLE
Lesson 9 - How to test for the WinnerPicked event

### DISCUSSION THREAD

**Discussion by pyromaus** [Cyfrin/foundry-full-course-cu#250] [General]
The `WinnerPicked` event takes the winner's address as a parameter. Foundry makes us emit the event before we pick the winner. How would we know who that winner is, if he's only picked later? Which address are we supposed to pass to the event?

The answer is any address will do. Just tell Foundry that you want it to ignore the address you pass to `WinnerPicked` .
On `vm.expectEmit` , pass all `false` like below. Make sure you pass `address(raffle)` as the emitter. Then emit `WinnerPicked` with any address you want. 

```solidity
function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public raffleEnteredAndTimePassed {
        ...
        ...
        vm.expectEmit(false, false, false, false, address(raffle));
        emit WinnerPicked(raffle.getPlayer(2));
        VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(uint(requestId), address(raffle));
        ...
        ...
}
```
Run the test with `-vvvv` : The first event below is emitted by our test - the address `(0x0.....002)` is from `raffle.getPlayer(2)` 
Then the real `WinnerPicked` is called on our Raffle contract.

```
emit WinnerPicked(winner: 0x0000000000000000000000000000000000000002)
    ├─ [20737] VRFCoordinatorV2Mock::fulfillRandomWords(1, Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1]) 
    │   ├─ [13396] Raffle::rawFulfillRandomWords(1, [78541660797044910968829902406342334108369226379826116161446442989268089806461 [7.854e76]]) 
    │   │   ├─ emit WinnerPicked(winner: 0x0000000000000000000000000000000000000001)
...
Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 8.49ms
```
The second event is the actual winner - a.k.a. `raffle.getPlayer(1)` . The addresses are different but the test passes. Leaving the first `false` as `true` on `vm.expectEmit` will make Foundry actually compare the two addresses, and your test will fail.

So is there a way Foundry can test for the emission of the event that includes the true winner? Idk. But for now, testing if `WinnerPicked` was emitted at all should suffice.

**Comment by ghost** [comment]
Hello @pyromaus , on a testnet or mainnet we can't beat the VRFCoordinatorV2 but locally we can try because it uses pseudo-randomness to compute the random words. This is how I did it:

```javascript
  function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public raffleEntered skipFork {
      // As in the repo enter the raffle with more players...

      // As in the repo collect the emitted event to get the requestId...

      uint256 REQUEST_NUMS = raffle.getNumWords();
      uint256[] memory words = new uint256[](REQUEST_NUMS);

      // cheat the VRFCoordinatorV2Mock by precomputing the random word
      for (uint256 i = 0; i < REQUEST_NUMS; i++) {
          words[i] = uint256(keccak256(abi.encode(requestId, i)));
      }

      // get the winner in advanced
      uint256 indexWinner = words[0] % raffle.getNumberOfPlayers();
      address expectedWinner = raffle.getPlayer(indexWinner);
      uint256 startingBalance = expectedWinner.balance;

      // kwoning the winner in advanced emit the event
      vm.expectEmit(true, false, false, false, address(raffle));
      emit WinnerPicked(expectedWinner);
      VRFCoordinatorV2Mock(vrfCoordinatorV2).fulfillRandomWords(uint256(requestId), address(raffle));

      // Assertions...
  }
```

**Reply by pyromaus** [reply]
True! Every time my test was repeated, the mock VRF output stayed exactly the same - makes total sense we can use this to compute the winner. Great work man

**Reply by sharonphiliplima** [reply]
Hey, is it really important to use this logic on our test to calculate the `winner` ?
I am just trying to check if an address was emitted...
The following doesn't pass:

```
vm.recordLogs();

        VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(
            uint256(requestId),
            address(raffle)
        );
        Vm.Log[] memory entry = vm.getRecordedLogs();

        entry = vm.getRecordedLogs();

        bytes32 winner = (entry[0].topics[1]);
        assert(address(uint160(uint256(winner))) == raffle.getRecentWinner());
    }
```

what am I doing wrong?

**Reply by ghost** [reply]
Yes, it's important because events are emitted and their arguments are stored in the transaction logs and in Foundry this is way to collect those args.

If I remember correctly two events are emitted:
- an event from the [VRFCoordinatorV2Mock](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol#L125), and the args of this event will be stored in the entry at index 0.
- the event from the Raffle after the winner is picked, which arg is stored in the entry at index 1.

So you should fetch the winner from `entry[1].topics[1]`

And you don't need `entry = vm.getRecordedLogs();`.

**Reply by sharonphiliplima** [reply]
Thanks, I removed the latter line. 
I don't understand why but it works with `entry[0].topics[1]` 

My test function:

```
function testFullfillRandomWordsCanOnlyBeCalledAfterPerformUpkeep(
        uint256 randomRequestId //(runs: 256, μ: 78317, ~: 78317) @timeoftesting
    ) public raffleEnteredAndTimePassed {
        //Act
        vm.expectRevert("nonexistent request");
        VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(
            randomRequestId,
            address(raffle)
        );
    }

    function testFullfillRandomWordsPickAWinnerResetsAndSendsMoney()
        public
        raffleEnteredAndTimePassed
    {
        //Arrange
        //function hoax(address who, uint256 give) public;  ~ prank+deal
        uint256 additionalEntrants = 5;
        uint256 startingIndex = 1; //there is already a player above!

        for (
            uint256 i = startingIndex;
            i < additionalEntrants + startingIndex;
            i++
        ) {
            address player = address(uint160(i));
            hoax(player, 1 ether);
            raffle.enterRaffle{value: entranceFee}();
        }

        //Act
        vm.recordLogs();
        raffle.performUpkeep(""); //emit requestId
        Vm.Log[] memory entries = vm.getRecordedLogs();
        bytes32 requestId = entries[1].topics[1];

        vm.recordLogs();
        VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(
            uint256(requestId),
            address(raffle)
        );
        Vm.Log[] memory entry = vm.getRecordedLogs();
        bytes32 winner = (entry[0].topics[1]);
        assert(address(uint160(uint256(winner))) == raffle.getRecentWinner());
        address W = address(uint160(uint256(winner)));
        console.log("winner:", W);
        console.log("recent winner:", raffle.getRecentWinner());
    }
}

```

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/69852010/c65a5eb6-396f-42c6-99f7-4ec0427ae478)

**Reply by ghost** [reply]
lol, for me it's weird. `vm.recordLogs` supposedly collects all the emitted events. But this time it only collected the event emitted by the Raffle.

I got curious and checked the repo but Patrick didn't use the cheatcode to fetch the winner.

What I can assume is that Foundry is only targeting the contracts deployed in `setUp`.

**Reply by sharonphiliplima** [reply]
I love your strategy for getting the winner in advance and then testing it. But would it be feasible always and most important, is it necessary?

**Reply by ghost** [reply]
It only work with contracts that use pseudo-randomness like the `VRFCoordinatorV2Mock`, but with contracts that actually use the VRFCoordinator it's useless because the randomness is computed off-chain. So I could not predict the winner of a Raffle deployed on a testnet or mainnet.

From a security standpoint this is important to know because depending on the application pseudo-randomness could be enough but in an application like a Raffle you need something more robust to prevent malicious users from exploiting the contract and stealing money.

**Reply by sharonphiliplima** [reply]
Now that you have mentioned security, I would love to learn the nuances more. Please do write a writeup in your free time or a twitter thread. Much thanks.
